native/src/main/java/io/ballerina/stdlib/data/jsondata/json/JsonTraverse.java

```diff
@@ -0,0 +1,309 @@
+            }
+            Map<String, Field> currentField = fieldHierarchy.pop();
+            checkOptionalFieldsAndLogError(currentField);
+            restType.pop();
+            return currentJsonNode;
+        }
+
+        private Object traverseArrayValue(BArray array, Object currentJsonNode) {
+            switch (rootArray.getTag()) {
+                case TypeTags.ARRAY_TAG -> {
+                    ArrayType arrayType = (ArrayType) rootArray;
+                    int expectedArraySize = arrayType.getSize();
+                    long sourceArraySize = array.getLength();
+                    if (expectedArraySize > sourceArraySize) {
+                        throw DiagnosticLog.error(DiagnosticErrorCode.ARRAY_SIZE_MISMATCH);
+                    }
+
+                    if (!allowDataProjection && expectedArraySize < sourceArraySize) {
+                        throw DiagnosticLog.error(DiagnosticErrorCode.ARRAY_SIZE_MISMATCH);
+                    }
+
+                    Type elementType = arrayType.getElementType();
+                    if (expectedArraySize == -1) {
+                        traverseArrayMembers(array.getLength(), array, elementType, currentJsonNode);
+                    } else {
+                        traverseArrayMembers(expectedArraySize, array, elementType, currentJsonNode);
+                    }
+                }
+                case TypeTags.TUPLE_TAG -> {
+                    TupleType tupleType = (TupleType) rootArray;
+                    Type restType = tupleType.getRestType();
+                    int expectedTupleTypeCount = tupleType.getTupleTypes().size();
+                    if (expectedTupleTypeCount > array.getLength()) {
+                        throw DiagnosticLog.error(DiagnosticErrorCode.ARRAY_SIZE_MISMATCH);
+                    }
+                    for (int i = 0; i < array.getLength(); i++) {
+                        Object jsonMember = array.get(i);
+                        Object nextJsonNode;
+                        if (i < expectedTupleTypeCount) {
+                            nextJsonNode = traverseJson(jsonMember, tupleType.getTupleTypes().get(i));
+                        } else if (restType != null) {
+                            nextJsonNode = traverseJson(jsonMember, restType);
+                        } else if (!allowDataProjection) {
+                            throw DiagnosticLog.error(DiagnosticErrorCode.ARRAY_SIZE_MISMATCH);
+                        } else {
+                            continue;
+                        }
+                        ((BArray) currentJsonNode).add(i, nextJsonNode);
+                    }
+                }
+            }
+            return currentJsonNode;
+        }
+
+        private void traverseArrayMembers(long length, BArray array, Type elementType, Object currentJsonNode) {
+            for (int i = 0; i < length; i++) {
+                ((BArray) currentJsonNode).add(i, traverseJson(array.get(i), elementType));
```

Comment:
In the `addRestField` methods we do a type compatible check before we adding values to the type.
But in here, we directly adding.
wdyt?

Reply:
Here also, we check with the member type when calling `traverseJson`.